File: frontend/src/services/auth.service.js
import { API_CONFIG, API_ENDPOINTS } from '../config/api.config';

/**
 * Authentication Service
 * Handles all authentication-related API calls through the API Gateway
 */

class AuthService {
  /**
   * Sign up a new user
   * @param {Object} userData - User registration data
   * @param {string} userData.userName - User's name
   * @param {string} userData.userEmail - User's email
   * @param {string} userData.userPassword - User's password
   * @param {string} userData.role - User's role (optional, defaults to "USER")
   * @returns {Promise<Object>} - API response
   */
  async signUp(userData) {
    try {
      const response = await fetch(
        `${API_CONFIG.USER_SERVICE}${API_ENDPOINTS.AUTH.SIGN_UP}`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            userName: userData.userName,
            userEmail: userData.userEmail,
            userPassword: userData.userPassword,
            role: userData.role || 'USER',
          }),
        }
      );

      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.message || 'Sign up failed');
      }

      return {
        success: true,
        data: data,
        message: data.message || 'User registered successfully',
      };
    } catch (error) {
      return {
        success: false,
        message: error.message || 'Something went wrong. Please try again.',
      };
    }
  }

  /**
   * Sign in a user
   * @param {Object} credentials - User login credentials
   * @param {string} credentials.userEmail - User's email
   * @param {string} credentials.userPassword - User's password
   * @returns {Promise<Object>} - API response with user data and tokens
   */
  async signIn(credentials) {
    try {
      const response = await fetch(
        `${API_CONFIG.USER_SERVICE}${API_ENDPOINTS.AUTH.SIGN_IN}`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
          },
          body: JSON.stringify({
            userEmail: credentials.userEmail,
            userPassword: credentials.userPassword,
          }),
        }
      );

      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.message || 'Sign in failed');
      }

      // Store tokens if sign in is successful
      if (data.body?.accessToken) {
        localStorage.setItem('accessToken', data.body.accessToken);
        localStorage.setItem('refreshToken', data.body.refreshToken);
        localStorage.setItem('user', JSON.stringify(data.body));
      }

      return {
        success: true,
        data: data.body,
        message: data.message || 'User signed in successfully',
      };
    } catch (error) {
      return {
        success: false,
        message: error.message || 'Something went wrong. Please try again.',
      };
    }
  }

  /**
   * Validate the current token
   * @returns {Promise<Object>} - Validation response
   */
  async validateToken() {
    try {
      const token = localStorage.getItem('accessToken');

      if (!token) {
        throw new Error('No token found');
      }

      const response = await fetch(
        `${API_CONFIG.USER_SERVICE}${API_ENDPOINTS.AUTH.VALIDATE_TOKEN}`,
        {
          method: 'POST',
          headers: {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${token}`,
          },
        }
      );

      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.message || 'Token validation failed');
      }

      return {
        success: true,
        data: data.body,
        message: data.message || 'Token is valid',
      };
    } catch (error) {
      // Clear invalid tokens
      this.clearTokens();
      return {
        success: false,
        message: error.message || 'Token validation failed',
      };
    }
  }

  /**
   * Get user by ID
   * @param {number} userId - User ID
   * @returns {Promise<Object>} - User data
   */
  async getUserById(userId) {
    try {
      const token = localStorage.getItem('accessToken');

      if (!token) {
        throw new Error('No token found');
      }

      const response = await fetch(
        `${API_CONFIG.USER_SERVICE}${API_ENDPOINTS.USER.GET_BY_ID(userId)}`,
        {
          method: 'GET',
          headers: {
            'Content-Type': 'application/json',
            Authorization: `Bearer ${token}`,
          },
        }
      );

      const data = await response.json();

      if (!response.ok) {
        throw new Error(data.message || 'Failed to fetch user');
      }

      return {
        success: true,
        data: data.body,
        message: data.message || 'User found',
      };
    } catch (error) {
      return {
        success: false,
        message: error.message || 'Failed to fetch user',
      };
    }
  }

  /**
   * Sign out the current user
   */
  signOut() {
    this.clearTokens();
  }

  /**
   * Clear all stored tokens and user data
   */
  clearTokens() {
    localStorage.removeItem('accessToken');
    localStorage.removeItem('refreshToken');
    localStorage.removeItem('user');
  }

  /**
   * Get the current user from localStorage
   * @returns {Object|null} - Current user data or null
   */
  getCurrentUser() {
    try {
      const userStr = localStorage.getItem('user');
      return userStr ? JSON.parse(userStr) : null;
    } catch {
      return null;
    }
  }

  /**
   * Get the current access token
   * @returns {string|null} - Access token or null
   */
  getAccessToken() {
    return localStorage.getItem('accessToken');
  }

  /**
   * Check if user is authenticated
   * @returns {boolean} - True if authenticated
   */
  isAuthenticated() {
    return !!this.getAccessToken();
  }
}

export default new AuthService();

File: UserService/src/main/java/com/sliit/userservice/service/AuthService.java
package com.sliit.userservice.service;

import com.sliit.userservice.dtos.UserData;
import com.sliit.userservice.dtos.UserResponse;
import com.sliit.userservice.dtos.requests.SignInRequest;
import com.sliit.userservice.dtos.requests.SignUpRequest;
import com.sliit.userservice.dtos.response.SignInResponse;
import com.sliit.userservice.entity.UserEntity;
import com.sliit.userservice.exception.InvalidCredentialsException;
import com.sliit.userservice.exception.UserAlreadyExistsException;
import com.sliit.userservice.exception.UserNotFoundException;
import com.sliit.userservice.repository.UserRepository;
import com.sliit.userservice.util.JwtTokenProvider;
import lombok.extern.slf4j.Slf4j;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.security.crypto.password.PasswordEncoder;
import org.springframework.stereotype.Service;

import java.util.Optional;

/**
 * @Author User-Service
 * @Created on 03/05/2025
 * @Project API-Gateway-Testing
 * @Description: Authentication Service for Sign Up, Sign In, and Token Validation
 */
@Slf4j
@Service
public class AuthService {

    @Autowired
    private UserRepository userRepository;

    @Autowired
    private JwtTokenProvider jwtTokenProvider;

    @Autowired
    private PasswordEncoder passwordEncoder;

    /**
     * Sign up new user
     */
    public UserResponse signUp(SignUpRequest request) {
        log.info("Sign up request for user: {}", request.getUserEmail());

        // Check if user already exists by email
        Optional<UserEntity> existingUserByEmail = userRepository.findByUserEmail(request.getUserEmail());
        if (existingUserByEmail.isPresent()) {
            log.warn("User already exists with email: {}", request.getUserEmail());
            throw new UserAlreadyExistsException("User already exists with this email");
        }

        // Check if username already exists
        Optional<UserEntity> existingUserByName = userRepository.findByUserName(request.getUserName());
        if (existingUserByName.isPresent()) {
            log.warn("Username already exists: {}", request.getUserName());
            throw new UserAlreadyExistsException("Username already exists");
        }

        // Create new user
        UserEntity newUser = UserEntity.builder()
                .userName(request.getUserName())
                .userEmail(request.getUserEmail())
                .userPassword(passwordEncoder.encode(request.getUserPassword()))
                .role(request.getRole() != null ? request.getRole() : "USER")
                .isActive(true)
                .build();

        UserEntity savedUser = userRepository.save(newUser);
        log.info("User registered successfully with email: {}", savedUser.getUserEmail());

        // Build response
        UserData userData = UserData.builder()
                .userId(savedUser.getUserId())
                .userEmail(savedUser.getUserEmail())
                .userName(savedUser.getUserName())
                .role(savedUser.getRole())
                .build();

        return UserResponse.builder()
                .message("User registered successfully")
                .statusCode(201)
                .data(userData)
                .build();
    }

    /**
     * Sign in user and return access token
     */
    public SignInResponse signIn(SignInRequest request) {
        log.info("Sign in request for user: {}", request.getUserEmail());

        // Find user by email
        UserEntity user = userRepository.findByUserEmail(request.getUserEmail())
                .orElseThrow(() -> {
                    log.warn("User not found with email: {}", request.getUserEmail());
                    return new UserNotFoundException("User not found with this email");
                });

        // Verify password
        if (!passwordEncoder.matches(request.getUserPassword(), user.getUserPassword())) {
            log.warn("Invalid password for user: {}", request.getUserEmail());
            throw new InvalidCredentialsException("Invalid email or password");
        }

        // Check if user is active
        if (!user.getIsActive()) {
            log.warn("User account is inactive: {}", request.getUserEmail());
            throw new UserNotFoundException("User account is inactive");
        }

        // Generate tokens
        String accessToken = jwtTokenProvider.generateToken(
                user.getUserId(),
                user.getUserEmail(),
                user.getUserName(),
                user.getRole()
        );

        String refreshToken = jwtTokenProvider.generateRefreshToken(user.getUserId());

        log.info("User signed in successfully: {}", request.getUserEmail());

        // Build response
        return SignInResponse.builder()
                .userId(user.getUserId())
                .userName(user.getUserName())
                .userEmail(user.getUserEmail())
                .role(user.getRole())
                .accessToken(accessToken)
                .refreshToken(refreshToken)
                .build();
    }

    /**
     * Validate access token and return user information
     */
    public UserResponse validateToken(String token) {
        log.info("Validating token");

        // Remove "Bearer " prefix if present
        if (token.startsWith("Bearer ")) {
            token = token.substring(7);
        }

        // Validate token
        if (!jwtTokenProvider.validateToken(token)) {
            log.warn("Invalid or expired token");
            throw new InvalidCredentialsException("Invalid or expired token");
        }

        // Extract user information from token
        Long userId = jwtTokenProvider.getUserIdFromToken(token);
        String userEmail = jwtTokenProvider.getUserEmailFromToken(token);
        String userName = jwtTokenProvider.getUserNameFromToken(token);
        String role = jwtTokenProvider.getRoleFromToken(token);

        log.info("Token validated successfully for user: {}", userEmail);

        // Build response
        UserData userData = UserData.builder()
                .userId(userId)
                .userEmail(userEmail)
                .userName(userName)
                .role(role)
                .build();

        return UserResponse.builder()
                .message("Token is valid")
                .statusCode(200)
                .data(userData)
                .build();
    }

    /**
     * Get user by ID (for internal service-to-service communication)
     */
    public UserResponse getUserById(Long userId) {
        log.info("Getting user by ID: {}", userId);

        UserEntity user = userRepository.findByUserIdAndIsActiveTrue(userId)
                .orElseThrow(() -> {
                    log.warn("User not found with ID: {}", userId);
                    return new UserNotFoundException("User not found with this ID");
                });

        UserData userData = UserData.builder()
                .userId(user.getUserId())
                .userEmail(user.getUserEmail())
                .userName(user.getUserName())
                .role(user.getRole())
                .build();

        return UserResponse.builder()
                .message("User found")
                .statusCode(200)
                .data(userData)
                .build();
    }
}

